config_type: add unstable_variant attribute

tommilligan opened the pull request:
First pass at #5378

Adds new functionality to the existing `config_proc_macro` crate, such that enums used for configuration can have their variants annotated as `unstable_variant`:

```rust
#[config_type]
/// How to merge imports.
pub enum ImportGranularity {
    /// Do not merge imports.
    Preserve,
    /// Use one `use` statement per crate.
    Crate,
    ...
    /// Adapt import granularity, based on internal heuristics.
    #[unstable_variant]
    Adaptive
}
```

This will be checked upon loading config, and a warning emitted similar to using current unstable options on stable toolchain:

```log
Warning: can't set `imports_granularity = Adaptive`, unstable variants are only available in nightly channel.
```

tommilligan commented on config_proc_macro/Cargo.lock:
```diff
@@ -1,68 +1,68 @@
 # This file is automatically @generated by Cargo.
 # It is not intended for manual editing.
+version = 3
```
I noted the lockfile versions have been upgraded by my local toolchain - happy to somehow keep to the original version if this is an issue for compatibility?

tommilligan reviewed (commented):
I have some questions for someone with more project experience

tommilligan commented:
I would also ask - how should I go about testing this? I couldn't find any tests for the existing unstable options, or the proc macro code internals. In fact, I get test failures when I run `cd config_proc_macro && cargo test` on master

calebcartwright commented:
Haven't really had a chance to look at this yet, but just wanted to say I love the idea! Attributes seem like a really clean and simple way of managing this and I'm excited about the possibilities

ytmimi commented:
@tommilligan as well as updating `fill_from_parsed_config` I also believe you'll need to modify `override_value`, which ends up getting called when overriding the config with any values passed on the command line using `--config`

ytmimi commented:
I can see how testing this could be a little tricky since we don't currently have any options that are stable with `unstable_variants`. I wonder if it would be possible to add a `config_type` to `Config` thats only available for testing.

```rust
// src/config/options.rs

/// This configuration option is only available for testing purposes
#[cfg(test)]
#[config_type]
pub enum SemiStableOption {
    Stable,
    #[unstable_variant]
    NotStable,
}
```

```rust
// src/config/config_types.rs

#[cfg(test)]
use crate::config::options::SemiStableOption;

macro_rules! create_config {
    ($($i:ident: $ty:ty, $def:expr, $stb:expr, $( $dstring:expr ),+ );+ $(;)*) => (
        #[cfg(test)]
        use std::collections::HashSet;
        use std::io::Write;

        use serde::{Deserialize, Serialize};

        #[derive(Clone)]
        #[allow(unreachable_pub)]
        pub struct Config {
            #[cfg(test)]
            semi_stable_option: (Cell<bool>, bool, SemiStableOption, bool),
            // For each config item, we store a bool indicating whether it has
            // been accessed and the value, and a bool whether the option was
            // manually initialised, or taken from the default,
            $($i: (Cell<bool>, bool, $ty, bool)),+
        }
...
```

And then we could write a few tests in `tests/rustfmt/main.rs` that call the rustfmt binary directly. You can take a look at some of the existing tests in that module. I also opened #5025 a while ago, which touched similar areas of the config code. It's in limbo right now, but some of the tests I wrote in [tests/rustmft/main.rs](https://github.com/rust-lang/rustfmt/pull/5025/files#diff-e62399b16396c027f430b956309ba52972d6930c505c5947295965999e916f20) could be useful to help you test that you get the expected warning messages when running on stable. You should be able to use the `#[stable_only_test]` attribute to only run the test when we set `CFG_RELEASE_CHANNEL=stable` to simulate running on a stable rustfmt.

Not 100% sure if this approach will work. This might cause some issues with the `configuration_snippet_tests`, which checks to make sure that each configuration option is documented so maybe we'll need to work around that? We can totally brainstorm some other options though!

tommilligan commented on config_proc_macro/src/attrs.rs:
```diff
@@ -1,8 +1,10 @@
+//!
+//! - `doc_hint`: name-value pair whose value is string literal
+//! - `value`: name-value pair whose value is string literal
+//! - `unstable_variant`: name only
```
I like this idea, but agree it should be in followup

tommilligan commented:
> Do we need to define a new `proc_macro_attribute` macro?

We don't - I've verified that the attribute works as intended locally. The reason is that the `unstable_variant` annotation is processed by the `config_type` proc macro and stripped off - so it's never actually 'called', if that makes sense.

> Also I believe the tests are failing because the doc tests for `nightly_only_test` and `stable_only_test` aren't quite right. We could fix them by adding `/// # use rustfmt_config_proc_macro::nightly_only_test;` and `/// # use rustfmt_config_proc_macro::stable_only_test;` to bring each attribute macro into scope for the doc tests. I can open up a PR for that a little later.

Thanks for this - I've opened #5389 with a fix for this plus CI to enforce it.

tommilligan commented:
> override_value

Added a test to show that updating `override_value` is not required, the value is parse from a JSON string as expected:

```rust
    #[test]
    fn test_override_skip_macro_names() {
        let mut config = Config::default();
        config.override_value("skip_macro_names", r#"["*", "println"]"#);
        assert_eq!(
            config.skip_macro_names(),
            MacroSelectors(vec![
                MacroSelector::All,
                MacroSelector::Name(MacroName::new("println".to_owned()))
            ])
        );
    }

```

tommilligan commented:
> override_value

Thanks for pointing this out. Updated to factor out the common logic of updating with an unstable option/value from `fill_from_parsed_config`, and pointed `override_value` to it as well.

I think this points out that you could set unstable options using `--config`, and this is _technically_ a breaking change that means you can't any more. However, I think this was never expected behaviour and should be classified as a bugfix.

tommilligan commented:
> I wonder if it would be possible to add a config_type to Config thats only available for testing.

Thanks for the hints on this. I actually found we already have a `mock` config object specifically for testing, so likewise added an option type for testing that has an unstable variant on it.

tommilligan commented:
@ytmimi Thanks for all your feedback - I think I've addressed everything you raised. CI appears to be failing with a lot of unrelated errors that I don't think are to do with this PR?

ytmimi commented:
> > @tommilligan as well as updating `Config::fill_from_parsed_config` I also believe you'll need to modify `Config::override_value`, which ends up getting called when overriding the config with any values passed on the command line using `--config`
> 
> Thanks for pointing this out. Updated to factor out the common logic of updating with an unstable option/value from `fill_from_parsed_config`, and pointed `override_value` to it as well.
> 
> I think this points out that you could set unstable options using `--config`, and this is _technically_ a breaking change that means you can't any more. However, I think this was never expected behaviour and should be classified as a bugfix.

If I had to guess I'd say it's this behavior that's causing the CI to fail, since it's only the "stable" tests that are failing. To reiterate, the "stable" tests don't actually run on a stable version of rustfmt they just run with `CFG_RELEASE_CHANNEL=stable` to simulate running on a stable rustfmt.

In theory I agree that this would be a bugfix for unintended behaviour, but I think there's probably a larger discussion that needs to be had around whether we want to continue to let users set unstable options via the command line. https://github.com/rust-lang/rustfmt/pull/5025 looked to solve this by giving users control of that decision, and allowing them to decide whether rustfmt exits with an abort message or warns and continues formatting when using unstable options.

For now can you simply issue the warning in `override_value` and still set the value regardless of the options stability. I think that will fix the issue you're seeing with CI

tommilligan commented:
> If I had to guess I'd say it's this behavior that's causing the CI to fail, since it's only the "stable" tests that are failing.

:facepalm: ah, thank you. Yes, makes-breaking-change-and-then-wonders-why-CI-is-failing sounds about right for me.

> For now, can you simply issue the warning in `override_value` and still set the value regardless of the options stability. I think that will fix the issue you're seeing with CI

I think it would be counterintuitive to warn about unstable options/variants, while at the same time allowing them to be set from `--config`. I think I will preserve the current behaviour, which is to allow setting any option/value, whether they are stable or not. I've added a comment to this effect, with a link to the relevant discussion PR that you posted.

If this isn't desired, I can look into changing the behaviour of `override_value`, but would like some design guidance on what the expected behaviour should be for unstable variants specifically.

Added unit tests that confirm the behaviour of both `fill_from_parsed_config` and `override_value` for unstable variants of options.

ytmimi commented:
> 🤦 ah, thank you. Yes, makes-breaking-change-and-then-wonders-why-CI-is-failing sounds about right for me.

Oh don't worry about it. I've definitely done the same thing before 😂

> I think it would be counterintuitive to warn about unstable options/variants, while at the same time allowing them to be set from --config. I think I will preserve the current behaviour, which is to allow setting any option/value, whether they are stable or not. I've added a comment to this effect, with a link to the relevant discussion PR that you posted.

That's fine. I think restoring the original behaviour for now makes sense and we can decide what the best way forward is after we've had some more time to think on it.

>Added unit tests that confirm the behaviour of both fill_from_parsed_config and override_value for unstable variants of options.

Fantastic! I took a look at the tests and everything looks good to me!

ytmimi commented on config_proc_macro/src/attrs.rs:
```diff
@@ -27,6 +29,11 @@ pub fn find_config_value(attrs: &[syn::Attribute]) -> Option<String> {
     attrs.iter().filter_map(config_value).next()
 }
 
+/// Returns `true` if the there is at least one `unstable` attribute in the given slice.
+pub fn find_unstable_variant(attrs: &[syn::Attribute]) -> bool {
+    attrs.iter().any(is_unstable_variant)
+}
```
A little bit of bike shedding on the name, but I feel like `has_unstable_variants` or `any_unstable_variants` would better convey that this returns a bool. Not a huge deal so don't feel like you need to change this.

tommilligan replied:
Renamed

ytmimi reviewed (approved):
Thanks for all your work on this! I feel pretty good about what we've got here. @calebcartwright I'm ready to move forward, but I definitely want to give you a chance to take a look before merging

calebcartwright commented:
This has unfortunately acquired a merge conflict, could you rebase and resolve when you get a chance? Overall I'm good with the approach, though may have a couple nits (e.g. function names) we can revisit at a later date.

> The only difference being made right now is that a warning will be issued if you set an unstable value on stable via the config file, but if you set it from the command line no warning will be issued. I think that's fine for now, but I'm personally in favor of issuing the warning even if we end up setting the value

We can continue that discussion on your PR to identify target behavior. The historical thinking, aiui, was to simply warn when the user asked rustfmt for an option that it wasn't going to apply. That's why the current behavior  when taking advantage of the `--config` loophole on stable doesn't produce any warnings (there's no need to warn the user an option they specified won't be used, because it's going to be used)

calebcartwright reviewed (approved):
Thank you so much for the excellent work on this!

It's a capability I've had in mind for a while but hadn't been able to make time for. I realize and appreciate your impetus was largely driven by `imports_granularity`, but I just want to stress this will have far greater and broader import around our ability to get more options out (at least partially) on stable :heart:

tommilligan commented:
Thanks, that's great to hear. Fingers crossed, look forward to seeing it in some good use cases in future! And thanks for the great iteration process from both of you, it's really made working on this feature a pleasure.
